src/tui/terminal.rs

```diff
@@ -0,0 +1,115 @@
+use crossterm::cursor;
+use crossterm::event;
+use crossterm::terminal;
+use miette::IntoDiagnostic;
+use miette::WrapErr;
+use ratatui::prelude::CrosstermBackend;
+use ratatui::prelude::Terminal;
+use std::io::Stdout;
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::panic;
+use std::sync::atomic::AtomicBool;
+use std::sync::atomic::Ordering;
+
+pub struct TerminalGuard {
+    terminal: Terminal<CrosstermBackend<Stdout>>,
+}
+
+impl Deref for TerminalGuard {
+    type Target = Terminal<CrosstermBackend<Stdout>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.terminal
+    }
+}
+
+impl DerefMut for TerminalGuard {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.terminal
+    }
+}
+
+impl Drop for TerminalGuard {
+    fn drop(&mut self) {
+        let result = exit().wrap_err("Failed to exit terminal during drop");
+        // Ignore the `Result` if we're already panicking; aborting is undesirable
+        if !std::thread::panicking() {
+            result.unwrap();
+        }
+    }
+}
+
+static INSIDE: AtomicBool = AtomicBool::new(false);
+
+pub fn enter() -> miette::Result<TerminalGuard> {
+    use event::KeyboardEnhancementFlags as KEF;
+
+    INSIDE.store(true, Ordering::SeqCst);
+
+    let mut stdout = std::io::stdout();
+
+    terminal::enable_raw_mode()
+        .into_diagnostic()
+        .wrap_err("Failed to enable raw mode")?;
+
+    crossterm::execute!(
+        stdout,
+        terminal::EnterAlternateScreen,
+        cursor::Hide,
+        event::EnableMouseCapture,
+        event::EnableFocusChange,
+        event::EnableBracketedPaste,
+        event::PushKeyboardEnhancementFlags(
+            KEF::DISAMBIGUATE_ESCAPE_CODES
+                | KEF::REPORT_EVENT_TYPES
+                | KEF::REPORT_ALL_KEYS_AS_ESCAPE_CODES
+        ),
+    )
```

Comment:
I could remove some of these, like focus change and bracketed paste events - but they don't make a big difference.

The spammy one that does make a dent is mouse events: did you know TUI apps can track mouse movement generally, like hover? So just moving the mouse over a TUI sends dozens of events. But we need mouse events for scrolling.